### PR TITLE
Added GitHub Action workflow to test all Python versions listed in the project config file.

### PR DESCRIPTION
**Pull request by nessita**
This works expands the Windows tests so they are run using every supported python version, which is taken from the project configuration file.

### CODE REVIEW

**Comment by felixxm** [comment]
Is it not a waste of resources and energy? I don't remember any Windows-specific regressions on older Python versions. (and we do this that way for many years). Even if there are any issues, daily builds will catch them.

**Comment by nessita** [comment]
Thank you for sharing your thoughts. I believe there are significant benefits to implementing this:

1. Aligning our testing approach across all supported Python versions, including Windows, ensures consistency with our practices on Linux. This proactive approach helps identify issues specific to particular Python versions early on, reducing the need for multiple contributor and maintainer cycles to iterate on PRs and follow up with fixes.
2. It's crucial for ensuring the correctness of our security patches. By integrating this matrix into our GitHub Actions workflow, we can catch test suite errors across all supported Python versions promptly, because during security patches development we can't rely on Jenkins. This enhancement is particularly valuable to the Fellows and would certainly remove a bit of the stress around building and releasing security patches.

Regarding concerns about resource use, I think it's a broader discussion, perhaps even a philosophical one, especially given today's global context where vast resources are allocated to activities that garner more attention.

Personally, I find it worthwhile to allocate a small additional amount of resources and energy (though considerably less than activities like bitcoin mining or data model training). This investment reduces the time spent on setting up multiple Python versions locally, running tests across stable branches, and ultimately alleviates stress for the Fellows. Additionally, it saves the DSF money by minimizing the need for extra efforts from our end.

**Comment by felixxm** [comment]
> Aligning our testing approach across all supported Python versions, including Windows, ensures consistency with our practices on Linux. This proactive approach helps identify issues specific to particular Python versions early on, reducing the need for multiple contributor and maintainer cycles to iterate on PRs and follow up with fixes.

Going that path we can add dozens of other test configurations e.g. for every support mix of GIS libraries, and spend hours on testing every PR.

> This enhancement is particularly valuable to the Fellows and would certainly remove a bit of the stress around building and releasing security patches.

For many years we didn't have any CI integrated into flow for security patches, I didn't find it particularly stressful, it's a checklist to do.

> Personally, I find it worthwhile to allocate a small additional amount of resources and energy (though considerably less than activities like bitcoin mining or data model training).

I would argue that finding a small inconsistency in tests (that has no consequences) once every few months/years is not worth checking all PRs for all Python versions on Windows. If you were inspired by

**Review by carltongibson — APPROVED** [review]
If @nessita and @sarahboyce want this change to help manage security releases, that’s just a straightforward :+1: — _Of course!_. 

The energy concern issue would be better handled by thinking about ways we can be smarter about which tests we run. For example, it’s been _several years_ since the idea of not running the full test suite for changes which only affect the docs was suggested, but I don’t think we’ve done anything about that. (Have we?) I think it’s a red-herring: do we spend CPU cycles for developer time — absolutely yes. 

(I wouldn’t block on it but, could the whole matrix be reserved for a manual trigger? Would that address the requirement?)

File: .github/workflows/tests.yml
Comment (felixxm) [inline .github/workflows/tests.yml:38]:
```suggestion
        python-version:
          - '3.10'
          - '3.11'
          - '3.12'
```

File: .github/workflows/tests.yml
Comment (felixxm) [inline .github/workflows/tests.yml:32]: If you really want to force it, why not to use the list of Python version (as we have in `schedule_tests.yml`) instead of complicated regexp?

Comment (nessita) [reply .github/workflows/tests.yml:32]: I wanted to provide a solution that would not need manual modification between `main` and (current, past and future) `stable` branches. Since each `stable` branch can potentially support different set of versions (specifically, `4.2.x` supports `3.8` and `3.9` which `5.0.x` does not support), calculating the matrix automatically from the project setup file felt the correct solution that both saves time not needing manual updates and also not having differences between the `main` and `stable`  branches.

A medium term goal that I also think would be beneficial is to split the Python version matrix building into a reusable component to re use it in multiple workflows if necessary.

Comment (felixxm) [reply .github/workflows/tests.yml:32]: I can see few issues here, it creates a barrier for many folks in understanding what is going on here. They are force to know regular expressions, `sed`, and to check in another file what kind of versions are supported. IMO, it's unnecessarily complicated.

Comment (nessita) [reply .github/workflows/tests.yml:32]: I disagree. The approach of defining the Python versions matrix independently of the branch simplifies maintenance by eliminating the need for manual updates in multiple files.

Considering that maintaining these files falls on the Fellows, I believe it's crucial to streamline processes to support our efforts effectively.

While I acknowledge the concern about complexity as a potential barrier, it's worth noting that these files typically don't receive frequent contributions from the community. Therefore, optimizing them for internal use can enhance efficiency without significant external impact.

**Comment by nessita** [comment]
@carltongibson I have added a second commit which reverts the initial proposal and adds a dedicated workflow file to run the Python Matrix as calculated from the project's config file, that can be triggered by either adding a label (label name suggestions welcome) or manually via `workflow_dispatch`. This new workflow would be particularly handy when dealing with security branches across all stable branches (my plan is to backport this work accordingly).

@sarahboyce we discussed this solution as a good alternative so I'm tagging you as well.

Thanks everyone!

**Review by carltongibson — APPROVED** [review]
@nessita Yep, great. 👍 

There were plenty of times I would have liked to see the whole matrix run for security patches _before_ merging them. 🙂

**Review by sarahboyce — APPROVED** [review]
Thank you

**Pull request merged**
